### ISSUE TITLE
sf::StencilBuffer not applied when drawn from RenderTexture to RenderWindow

### ISSUE THREAD

**Issue by danieljpetersen** [SFML/SFML#3107]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I am here to report an issue and not to just ask a question or look for help (use the [forum](https://en.sfml-dev.org/forums/index.php#c3) or [Discord](https://discord.gg/nr4X7Fh) instead)

### Describe your issue here

sf::StencilBuffer appears to not work when drawn onto an sf::RenderTexture which is then drawn to an sf::RenderWindow. 

It behaves correctly when drawn directly onto an sf::RenderWindow. The issue is strictly sf::RenderTexture -> sf::RenderWindow.

It behaves correctly when drawn onto an sf::RenderTexture which is then copied to an sf::Image which explains why we are not getting failing tests.

I don't understand all the intricacies of SFML's OpenGL implementation (or OpenGL for that matter) so I have limited ability to dive into this. I did find that the issue is fixed if we force a resetGLStates() first thing within RenderTarget::setupDraw(bool useVertexCache, const RenderStates& states).

https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/RenderTarget.cpp#L833

To me this implies that the issue may be with caching the state, but this is just a shot in the dark.

### Your Environment

- OS / distro / window manager: Linux / Fedora 40 / Gnome 46.2
- SFML version: The latest master, d99d673330fa1b3d37d3dc09a147c6ad45df2c1b
- Compiler / toolchain: gcc 14.1.1

### Steps to reproduce

This is the [official StencilBuffer example](https://github.com/SFML/SFML/blob/master/examples/stencil/Stencil.cpp) minimally modified to draw onto an intermediate RenderTexture

```cpp
#include <SFML/Graphics.hpp>
#include <cstdlib>

int main()
{
    auto contextSettings = sf::ContextSettings{0 /* depthBits */, 8 /* stencilBits */};

    sf::RenderWindow window(sf::VideoMode({600, 600}),
                            "SFML Stencil",
                            sf::Style::Titlebar | sf::Style::Close,
                            sf::State::Windowed,
                            contextSettings);
    window.setVerticalSyncEnabled(true);

    sf::RenderTexture renderTexture = sf::RenderTexture::create(sf::Vector2u(600, 600), contextSettings).value();

    sf::RectangleShape red({500, 50});
    red.setFillColor(sf::Color::Red);
    red.setPosition({270, 70});
    red.setRotation(sf::degrees(60));

    sf::RectangleShape green({500, 50});
    green.setFillColor(sf::Color::Green);
    green.setPosition({370, 100});
    green.setRotation(sf::degrees(120));

    sf::RectangleShape blue({500, 50});
    blue.setFillColor(sf::Color::Blue);
    blue.setPosition({550, 470});
    blue.setRotation(sf::degrees(180));

    while (window.isOpen())
    {
        while (const auto event = window.pollEvent())
        {
            if (event.is<sf::Event::Closed>())
            {
                window.close();
                break;
            }
        }

    	window.clear(sf::Color::Black, 0);
    	renderTexture.clear(sf::Color::Black, 0);

        renderTexture.draw(red,
                    sf::StencilMode{sf::StencilComparison::Always, sf::StencilUpdateOperation::Replace, 3, ~0u, false});

        renderTexture.draw(green,
                    sf::StencilMode{sf::StencilComparison::Always, sf::StencilUpdateOperation::Replace, 1, ~0u, false});

        renderTexture.draw(blue,
                    sf::StencilMode{sf::StencilComparison::Greater, sf::StencilUpdateOperation::Replace, 2, ~0u, false});

        renderTexture.display();
        window.draw(sf::Sprite(renderTexture.getTexture()));
        window.display();
    }

    return EXIT_SUCCESS;
}
```

### Expected behavior

![expected](https://github.com/SFML/SFML/assets/5576311/35eaa4ae-a96a-49f4-9901-efc2cede8ba9)

### Actual behavior

![actual](https://github.com/SFML/SFML/assets/5576311/03d07f7e-6490-41ec-84c6-a44408d211cb)

**Comment by vittorioromeo** [comment]
Doing a bit of investigation. Instead of calling `resetGLStates` in `setupDraw`, it is sufficient to call `m_cache.stencilEnabled = false;` to fix the issue:

    if (!m_cache.glStatesSet)
        resetGLStates();

    m_cache.stencilEnabled = false; // <---

Also, calling `renderTexture.setActive(false)` before drawing the shapes fixes the issue, probably because `m_cache.enable` is set to `false`.

**Comment by danieljpetersen** [comment]
It also appears to work correctly for the first frame drawn, which is also probably because m_cache.enable is set to false.

**Comment by vittorioromeo** [comment]
> It also appears to work correctly for the first frame drawn. I'm assuming this is because m_cache.glStatesSet would be false for that frame (haven't verified this).

Creating the `RenderTexture` every frame inside the loop from scratch also fixes the issue.

**Comment by vittorioromeo** [comment]
I think I understand the problem:

- The first call to `resetGLStates` for the render texture calls `glDisable(GL_STENCIL_TEST)`.
- Then, the subsequent call to `applyStencilMode` calls `glEnable(GL_STENCIL_TEST)`.
- Then, we get to `window.draw(sf::Sprite(renderTexture.getTexture()))`.
- The `setupDraw` call for the sprite draw calls `resetGLStates`, which sets  `glDisable(GL_STENCIL_TEST)`.
- `glEnable(GL_STENCIL_TEST)` is never called again.

Adding `glEnable(GL_STENCIL_TEST)` right after `window.draw(sf::Sprite(renderTexture.getTexture()))` fixes the issue.

**Comment by vittorioromeo** [comment]
Another possible fix is doing this in `applyStencilMode`:

    // Apply the stencil mode
    if (!m_cache.stencilEnabled || !glIsEnabled(GL_STENCIL_TEST))
    
I don't know what the "correct" fix is. @binary1248?

**Comment by danieljpetersen** [comment]
> Adding `glEnable(GL_STENCIL_TEST)` right after `window.draw(sf::Sprite(renderTexture.getTexture()))` fixes the issue.

> if (!m_cache.stencilEnabled || !glIsEnabled(GL_STENCIL_TEST))

It's possible I'm just doing something wrong somewhere, but I don't think either of these fix the issue in all cases. Some context first...

The reason I found this bug is because I had implemented veronoi split screen in the past by using two sf::RenderTextures. I rendered the scene twice, one for each render texture, and then rendered the relevant portions from each texture onto the window.

I'm now rendering the scene onto a single render texture and using StencilBuffers to only render the relevant portion.

I attempted to use both of the proposed fixes in my project (glEnable(GL_STENCIL_TEST) in usercode as well as inserting the call after // Apply the stencil mode) and neither worked. 

It does fix it for the simple example case posted in the initial issue though. For my project, always calling resetGLStates() in setupDraw fixes it. Obviously this isn't a suitable fix though.

I can try to extract what my project is doing into a complete minimal example if people think it would be useful but I'd rather not if it won't actually be useful.

At a high level,

```cpp
// defining the part of the screen we want content to be visible
sf::RenderStates stencilState;
stencilState.stencilMode.stencilComparison = sf::StencilComparison::Always;
stencilState.stencilMode.stencilUpdateOperation = sf::StencilUpdateOperation::Replace;
stencilState.stencilMode.stencilReference = stencilValue;
stencilState.stencilMode.stencilMask = 0xFF;
stencilState.stencilMode.stencilOnly = true;

renderTexture.draw(polygon, stencilState);

// ...
// drawing the content
sf::RenderStates player1States = states;
player1States.stencilMode.stencilComparison = sf::StencilComparison::Equal;
player1States.stencilMode.stencilReference = sf::StencilValue(1);
player1States.stencilMode.stencilMask = 0xFF;
renderTexture.draw(drawable, player1States);

```

**Comment by danieljpetersen** [comment]
I felt it was lame of me to not provide a full example in my last response. I'm sorry about that.

This example will work if you call resetGLStates, but glEnable(GL_STENCIL_TEST) is not sufficient. 

```cpp
#include <SFML/OpenGL.hpp>
#include <SFML/Graphics.hpp>
#include <vector>
#include <iostream>

void applyStencilForLeftHalfScreen(sf::RenderTexture& renderTexture)
{
    std::vector<sf::Vector2f> points = {
        {0, 0},
        {300.0f, 0},
        {300.0f, 600},
        {0, 600}
    };

    sf::VertexArray polygon(sf::PrimitiveType::TriangleFan);
    for (const auto& point : points)
    {
        polygon.append(sf::Vertex(point));
    }

    sf::RenderStates stencilState;
    stencilState.stencilMode.stencilComparison = sf::StencilComparison::Always;
    stencilState.stencilMode.stencilUpdateOperation = sf::StencilUpdateOperation::Replace;
    stencilState.stencilMode.stencilReference = 1;
    stencilState.stencilMode.stencilMask = 0xFF;
    stencilState.stencilMode.stencilOnly = true;

    renderTexture.draw(polygon, stencilState);
}

int main()
{
    auto contextSettings = sf::ContextSettings{0 /* depthBits */, 8 /* stencilBits */};

    sf::RenderWindow window(sf::VideoMode({600, 600}),
                            "SFML Stencil",
                            sf::Style::Titlebar | sf::Style::Close,
                            sf::State::Windowed,
                            contextSettings);
    window.setVerticalSyncEnabled(true);

    sf::RenderTexture renderTexture = sf::RenderTexture::create(sf::Vector2u(600, 600), contextSettings).value();

    sf::RectangleShape fullScreenShape({600, 600});
    fullScreenShape.setFillColor(sf::Color::Yellow);

    while (window.isOpen())
    {
        while (const auto event = window.pollEvent())
        {
            if (event.is<sf::Event::Closed>())
            {
                window.close();
                break;
            }
        }

        window.clear(sf::Color::Black, 0);
        renderTexture.clear(sf::Color::Black, 0);

        // Apply stencil for left half of the screen
        applyStencilForLeftHalfScreen(renderTexture);

        // Draw a full-screen rectangle but only the left half should be visible due to the stencil
        sf::RenderStates leftStates;
        leftStates.stencilMode.stencilComparison = sf::StencilComparison::Equal;
        leftStates.stencilMode.stencilReference = 1;
        leftStates.stencilMode.stencilMask = 0xFF;
        renderTexture.draw(fullScreenShape, leftStates);

        renderTexture.display();

        // Draw render texture to window
        window.draw(sf::Sprite(renderTexture.getTexture()));

        // Doesn't fix the issue
        glEnable(GL_STENCIL_TEST);

        window.display();
    }

    return EXIT_SUCCESS;
}
```

# Expected
![expected](https://github.com/SFML/SFML/assets/5576311/f24dd4f3-14d3-4eb0-8c6b-94c6de02c04c)

# Actual 
![actual](https://github.com/SFML/SFML/assets/5576311/75fab77f-5019-4f22-b713-8ec22c4e3754)

**Comment by binary1248** [comment]
Should be fixed in #3114.
